drivers/flash/flash_stm32.c

```diff
@@ -6,10 +6,29 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#define DT_DRV_COMPAT soc_nv_flash
-
 #include <kernel.h>
 #include <device.h>
+
+#if DT_HAS_NODE(DT_INST(0, st_stm32f0_flash_controller))
+#define DT_DRV_COMPAT st_stm32f0_flash_controller
+#elif DT_HAS_NODE(DT_INST(0, st_stm32f1_flash_controller))
+#define DT_DRV_COMPAT st_stm32f1_flash_controller
+#elif DT_HAS_NODE(DT_INST(0, st_stm32f3_flash_controller))
+#define DT_DRV_COMPAT st_stm32f3_flash_controller
+#elif DT_HAS_NODE(DT_INST(0, st_stm32f4_flash_controller))
+#define DT_DRV_COMPAT st_stm32f4_flash_controller
+#elif DT_HAS_NODE(DT_INST(0, st_stm32f7_flash_controller))
+#define DT_DRV_COMPAT st_stm32f7_flash_controller
+#elif DT_HAS_NODE(DT_INST(0, st_stm32g0_flash_controller))
+#define DT_DRV_COMPAT st_stm32g0_flash_controller
+#elif DT_HAS_NODE(DT_INST(0, st_stm32g4_flash_controller))
+#define DT_DRV_COMPAT st_stm32g4_flash_controller
+#elif DT_HAS_NODE(DT_INST(0, st_stm32l4_flash_controller))
+#define DT_DRV_COMPAT st_stm32l4_flash_controller
+#elif DT_HAS_NODE(DT_INST(0, st_stm32wb_flash_controller))
+#define DT_DRV_COMPAT st_stm32wb_flash_controller
+#endif /* DT_HAS_NODE(DT_INST(0, st_stm32xx_flash_controller)) */
```

Comment:
Would be nice to use a STM32 generic compatible along with the soc series specific ones to avoid this.
I've submitted https://github.com/zephyrproject-rtos/zephyr/issues/24113 for this purpose